packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -94,356 +94,172 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
         bytes32 stateHash,
         uint256[] memory indices
     ) external override {
-        // checks
-        _requireIncreasingIndices(indices); // This assumption relied on by _computeNewAllocation (below)
-        _requireChannelFinalized(fromChannelId);
-        _requireMatchingFingerprint(stateHash, keccak256(outcomeBytes), fromChannelId);
-
-        Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
-        Outcome.AssetOutcome memory assetOutcome = abi.decode(
-            outcome[assetIndex].assetOutcomeBytes,
-            (Outcome.AssetOutcome)
-        );
-        require(
-            assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
-            '!allocation'
-        );
-        Outcome.AllocationItem[] memory allocation = abi.decode(
-            assetOutcome.allocationOrGuaranteeBytes,
-            (Outcome.AllocationItem[])
-        );
-        address asset = outcome[assetIndex].asset;
-
-        // ************************
-        // effects and interactions
-        // ************************
-        uint256 initialHoldings;
-
-        // execute the exit at the specified indices, updating allocation in place to the updated
-        // allocation returned by _transfer
-        (allocation, initialHoldings) = _transfer(asset, fromChannelId, allocation, indices);
+        (
+            Outcome.SingleAssetExit[] memory outcome,
+            address asset,
+            uint256 initialAssetHoldings
+        ) = _apply_transfer_checks(assetIndex, indices, fromChannelId, stateHash, outcomeBytes); // view
 
-        // Update the fingerprint
-        outcome[assetIndex].assetOutcomeBytes = abi.encode(
-            Outcome.AssetOutcome(Outcome.AssetOutcomeType.Allocation, abi.encode(allocation))
+        (
+            Outcome.Allocation[] memory newAllocations,
+            ,
```

Comment:
Any particular reason to keep this parameter?

Reply:
Gotcha! Is there a case where a single asset is being transferred, but it is the last asset to be transferred? Then would the entire outcome allocate zero amounts?

Reply:
In principle yes. But since a) we don't remove any parts of the outcome during a partial exit and b) we don't tend to inspect other parts of the outcome besides the assets targeted by the blockchain transaction, it's still a bit awkward to detect that it's safe to clear out the storage slot. 

We might want to factor this logic out in to a `deleteStatusIfOutcomeIsEmpty` helper. But I expect that may be somewhat expensive gas wise.

Reply:
Caveat: I have not read the code, only this conversation. (On my phone, making code review difficult.)

 Why don’t we remove any parts of the outcome during a partial exit? 

For transfer, it is necessary to remove parts of the outcome to prevent the first payee from being able to drain the funds.

For claim, it still seems like the right thing to do, rather than depend on a correct implementation of the virtual funding protocol to ensure that subsequent claims have no effect, right?

Reply:
We don't remove parts of the outcome in `transfer` either. We simply set the relevant `amount` to zero. 

Actually changing the shape of the outcome is tricky because of the way arrays work in solidity. Furthermore, things like `assetIndex` can have a dynamic meaning if the shape of the outcome can change. 

What is relatively straightforward is to check when all the amounts are zero, and then delete the entire outcome if safe to do so.

Reply:
Got it. Setting the amount to 0 makes sense!